### ISSUE TITLE
Unify 'tests' target for ct, eunit, triq, ...

### ISSUE THREAD

**Issue by zuiderkwast** [ninenines/erlang.mk#200]
I know that you @essen are working on this already but I think it could be nice to have an issue where people can add their ideas. You could assign yourself to it so that people can see what's going on. Here are a few of my ideas:

* Rename `build-ct-deps` to `test-deps` and make `test-deps` a prerequisite of `tests` rather than of `build-ct-suites`.
* ct shouldn't fail if there is no `_SUITE` in `test/` (as noted in #195).
* Coverage report using `cover`, preferably for eunit, ct and other kinds of tests, which would require all these to run in the same vm instance. Pseudo code:
  `cover:enable("src/*.erl"), eunit:run($(EUNIT_MODULES)), ct:run($(CT_SUITES)), triq:check($(TRIQ_MODULES)), $(USER_DEFINED_TEST_CODE), cover:report().`
* Make it possible to hook in other unsupported testing methods e.g. proper or quviq eqc using some makefile variable such as `$(USER_DEFINED_TEST_CODE)` above.

**Comment by essen** [comment]
First is done and will be pushed today (with a twist, there's a test-build target that eunit/ct both depend on, this is to enable parallel testing).

Second point, noted, I will try that now.

Cover is out of scope at this point though I am not forgetting, and still think we should go with https://github.com/ramsay-t/Smother instead of the less powerful cover. But either way it should be pluggable so that people who really want cover can later add it.

I will be adding the triq plugin also today. Not sure what you mean with the makefile variable, you can already add your own testing methods the same way eunit was added. Just add e.g. `tests:: proper` line and write the `proper` target.

The bigger change I introduce here is the concept of a "test build" to avoid rebuilding everything for every test tool. Seems to work great in my testing.

**Comment by zuiderkwast** [comment]
Nice. What I meant by the makefile variable is that I'd like a way to run ct and eunit and e.g. proper tests with a combined coverage report for all these testruns. That's what i tried to illustrate with the pseudo code anyway. If there is another way to run them separately and merge separate the cover reports afterwards, that would be fine too. Now I'll have a look at Smother.

**Comment by essen** [comment]
Right this is preferable if it is possible.

**Comment by essen** [comment]
Another thing I want to do is introduce the "t" variable the same way I use it with triq:

$(t) isn't defined - run everything
$(t) is defined and doesn't contain ":" - run the module $(t)
$(t) is defined and contains ":" - run the test case function $(t)/0

This should work for all of ct, eunit and triq, but if later they get added also proper or eqc.

It won't be part of the push I am finalizing now though. (Takes me forever, I know.)

**Comment by essen** [comment]
Here goes nothing! 7ee4e7958acf0f0956bd64023ecb1731b2824b5e

**Comment by zuiderkwast** [comment]
Nice refactoring work! (And my doc-deps branch can still be merged cleanly.)

Regarding merging multiple coverage reports, it appears to be possible with `cover` using export/1,2 and import/1. See http://www.erlang.org/doc/man/cover.html#export-1. Smother doesn't seem to have this ability (yet).

**Comment by essen** [comment]
Right I forgot to update copyright year, so it'll probably break after I do that but I can do it after merging yours. Will only look tomorrow though, I want triq in tonight and then going to play games.

Good to know about cover. Not sure how important coverage is for people so not sure what priority that should get. Otherwise I have a number of things I want to add to erlang.mk: kerl/ci thing (see Cowboy's all.sh), autopatch to get rid of build issues and smelly code grep (only have a handful of cases for now).

**Comment by zuiderkwast** [comment]
I don't know about other people but coverage is important to the projects that I'm involved in. We'll most likely stick with rebar for tests until we have coverage. Why don't you create issues for all your ideas so that people can comment on them?

**Comment by essen** [comment]
I mentioned autopatch as a possible solution in a couple open issues. The smelly code grep is not coming soon, I am experimenting with it against OTP and if it proves useful will add it to erlang.mk. The kerl/ci thing is just converting a shell script to erlang.mk, there's not much to comment on except whether or not to control kerl from erlang.mk or not (and the answer is probably yes).

**Comment by zuiderkwast** [comment]
No only that if people are interested in a certain feature they tend to comment "+1" and such things.

**Comment by essen** [comment]
Sure but for things I don't use much myself I want patches more than +1s. :-) I haven't really needed tickets to evaluate priority, talking with people is enough. Right now I would say cover to be the next thing people want most, alongside autopatch (or rather "something that fixes broken deps easily"), the latter fixing actual issues so I will focus on that next.

**Comment by zuiderkwast** [comment]
I can implement cover as long as we agree more or less how it should be done.

**Comment by essen** [comment]
Sure, at this point other than "we want all tests to get the same report" I got nothing. Will require some thinking to figure that one out too from the looks of it. :-)

**Comment by essen** [comment]
Autopatch: https://github.com/ninenines/erlang.mk/commit/1729811327320b0f64e54331d3f469399c6898da

**Comment by zuiderkwast** [comment]
Shouldn't we check that triq exists before using it? Now triq is required by default for `make tests`. Here is how to reproduce:

```
$ wget https://github.com/ninenines/erlang.mk/raw/master/erlang.mk
$ make -f erlang.mk bootstrap-lib
$ make tests
 GEN    clean-app
 APP    x.app.src
 GEN    test-dir
 GEN    test-build
find: "test": Filen eller katalogen finns inte
 GEN    eunit
======================== EUnit ========================
directory "ebin"
  There were no tests to run.
 GEN    triq
make: *** [triq] Error 1
```

Another thing: In triq_run there is a `catch error:undef` clause that does `halt()` which I think it should be `halt(1)`.

**Comment by zuiderkwast** [comment]
Hm... actually, that's not why it fails. It fails because what is checked when there are no modules to test is `[true] =:= lists:usort([triq:check(M) || M <- []]` which is false, thus `halt(1)`. I think that comparison should be something like `lists:all(fun triq:check/1, [$(TRIQ_MODULES)])` instead.

I realize now that `catch error:undef` is for the case when triq doesn't exist. I think this should be silently ignored.

**Comment by essen** [comment]
The error:undef is for when you ask for a specific test module or property using the t variable. The test for whether triq exists should be done before that, by calling module_info for example.

**Comment by essen** [comment]
Triq now no longer runs if deps/triq does not exist.

**Comment by essen** [comment]
Also fixed an error message caused by the eunit plugin when there is no test/ folder.

**Comment by essen** [comment]
I believe everything here is covered. Coverage itself has a separate PR that I will get to either today (unlikely) or this week-end/early next week, so I will be closing this ticket. Thanks!
